File: src/apps/jhonny_home_studio_app/lib/features/auth/data/auth_service.dart
import '../../../core/errors/api_exception.dart';
import '../../../core/network/api_client.dart';
import '../../../core/storage/token_storage.dart';
import 'auth_models.dart';

class AuthService {
  AuthService({
    required ApiClient apiClient,
    required TokenStorage tokenStorage,
  }) : _apiClient = apiClient,
       _tokenStorage = tokenStorage;

  final ApiClient _apiClient;
  final TokenStorage _tokenStorage;

  Future<AuthUser> login(String email, String password) async {
    final response = await _apiClient.postJson(
      '/auth/login',
      data: LoginRequest(email: email, password: password).toJson(),
    );
    return _saveAndBuildUser(response);
  }

  Future<AuthUser> register(RegisterCustomerRequest request) async {
    final response = await _apiClient.postJson(
      '/auth/register-customer',
      data: request.toJson(),
    );
    return _saveAndBuildUser(response);
  }

  Future<void> logout() async {
    await _tokenStorage.deleteToken();
  }

  Future<String?> getToken() async {
    return _tokenStorage.getToken();
  }

  Future<bool> isLoggedIn() async {
    final token = await _tokenStorage.getToken();
    return token != null && token.isNotEmpty;
  }

  Future<AuthUser> _saveAndBuildUser(Map<String, dynamic> response) async {
    final data = response['data'];
    if (data is! Map<String, dynamic>) {
      throw ApiException(message: 'Dados de autenticação inválidos.');
    }

    final user = AuthUser.fromJson(data);
    if (user.token.isEmpty) {
      throw ApiException(message: 'Token de autenticação não encontrado.');
    }

    await _tokenStorage.saveToken(user.token);
    return user;
  }
}

File: src/apps/jhonny_home_studio_admin/src/app/core/services/auth.service.ts
import { HttpClient, HttpErrorResponse } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { catchError, map, Observable, tap, throwError } from 'rxjs';

import { ApiConfig } from '../config/api-config';
import { ApiResponse } from '../models/api-response.model';
import { AuthUser, LoginRequest, RegisterRequest } from '../models/auth.model';
import { TokenService } from './token.service';

@Injectable({ providedIn: 'root' })
export class AuthService {
  private readonly loginUrl = `${ApiConfig.baseUrl}/api/auth/login`;
  private readonly registerUrl = `${ApiConfig.baseUrl}/api/auth/register-customer`;

  constructor(
    private readonly http: HttpClient,
    private readonly tokenService: TokenService
  ) {}

  login(request: LoginRequest): Observable<AuthUser> {
    return this.http.post<ApiResponse<AuthUser>>(this.loginUrl, request).pipe(
      map((response) => this.readAuthResponse(response)),
      tap((user) => this.persistSession(user)),
      catchError((error: unknown) => throwError(() => new Error(this.readError(error))))
    );
  }

  register(request: RegisterRequest): Observable<AuthUser> {
    return this.http.post<ApiResponse<AuthUser>>(this.registerUrl, request).pipe(
      map((response) => this.readAuthResponse(response)),
      tap((user) => this.persistSession(user)),
      catchError((error: unknown) => throwError(() => new Error(this.readError(error))))
    );
  }

  logout(): void {
    this.tokenService.clear();
  }

  isAuthenticated(): boolean {
    return this.tokenService.hasToken() && this.tokenService.getUser() !== null;
  }

  currentUser(): AuthUser | null {
    return this.tokenService.getUser();
  }

  private persistSession(user: AuthUser): void {
    this.tokenService.saveToken(user.token);
    this.tokenService.saveUser(user);
  }

  private readAuthResponse(response: ApiResponse<AuthUser>): AuthUser {
    if (!response.success) {
      throw new Error(this.readMessage(response));
    }

    const user = response.data;

    if (!user.token) {
      throw new Error('Token de autenticação indisponível.');
    }

    return user;
  }

  private readError(error: unknown): string {
    if (error instanceof HttpErrorResponse) {
      if (error.status === 0) {
        return 'Não foi possível conectar à API. Verifique o backend.';
      }

      return this.readMessage(error.error as Partial<ApiResponse<unknown>> | null) || 'Não foi possível entrar.';
    }

    if (error instanceof Error) {
      return error.message;
    }

    return 'Não foi possível entrar.';
  }

  private readMessage(response: Partial<ApiResponse<unknown>> | null): string {
    if (!response) {
      return '';
    }

    return response.errors?.length ? response.errors.join(' ') : response.message ?? '';
  }
}
